src/config/compiler.rs

```diff
@@ -95,26 +95,29 @@ fn expand_wildcards(config: &mut ConfigBuilder) {
         .collect::<Vec<String>>();
 
     for (name, transform) in config.transforms.iter_mut() {
-        expand_wildcards_inner(&mut transform.inputs, name, &candidates);
+        expand_globs_inner(&mut transform.inputs, name, &candidates);
     }
 
     for (name, sink) in config.sinks.iter_mut() {
-        expand_wildcards_inner(&mut sink.inputs, name, &candidates);
+        expand_globs_inner(&mut sink.inputs, name, &candidates);
     }
 }
 
-fn expand_wildcards_inner(inputs: &mut Vec<String>, name: &str, candidates: &[String]) {
+fn expand_globs_inner(inputs: &mut Vec<String>, name: &str, candidates: &[String]) {
     let raw_inputs = std::mem::take(inputs);
     for raw_input in raw_inputs {
-        if raw_input.ends_with('*') {
-            let prefix = &raw_input[0..raw_input.len() - 1];
-            for input in candidates {
-                if input.starts_with(prefix) && input != name {
-                    inputs.push(input.clone())
+        match glob::Pattern::new(&raw_input) {
+            Ok(pattern) => {
+                for input in candidates {
+                    if pattern.matches(input) && input != name {
+                        inputs.push(input.clone())
+                    }
                 }
             }
-        } else {
-            inputs.push(raw_input);
+            Err(error) => {
+                error!(message = "Invalid glob pattern for input.", component_name = name, %error);
+                continue;
+            }
```

Comment:
Should this error, or just do an exact match?

Reply:
Hmm, that's true, maybe an exact match would be acceptable. Although the user may be confused if they were trying to use a pattern that ended up being invalid.

Maybe we could emit a warning and then do an exact match?

Reply:
That sounds best, yes.

Reply:
👍 updated in https://github.com/timberio/vector/pull/7873/commits/53441839133a229dc7c4fbcc810bf0de04532a46